dissect/target/volumes/luks.py

```diff
@@ -0,0 +1,116 @@
+import logging
+import pathlib
+from typing import BinaryIO, Iterator, Optional, Union
+
+from dissect.fve import luks
+from dissect.util.stream import AlignedStream
+
+from dissect.target.exceptions import VolumeSystemError
+from dissect.target.helpers.keychain import KeyType
+from dissect.target.volume import EncryptedVolumeSystem, Volume
+
+log = logging.getLogger(__name__)
+
+
+class LUKSVolumeSystemError(VolumeSystemError):
+    pass
+
+
+class LUKSVolumeSystem(EncryptedVolumeSystem):
+    PROVIDER = "luks"
+
+    def __init__(self, fh: Union[BinaryIO, list[BinaryIO]], *args, **kwargs):
+        super().__init__(fh, *args, **kwargs)
+        self.luks = luks.LUKS(fh)
+
+    @staticmethod
+    def _detect(fh: BinaryIO) -> bool:
+        return luks.is_luks_volume(fh)
+
+    def _volumes(self) -> Iterator[Volume]:
+        if isinstance(self.fh, Volume):
+            volume_details = dict(
+                number=self.fh.number,
+                offset=self.fh.offset,
+                vtype=self.fh.type,
+                name=self.fh.name,
+            )
+        else:
+            volume_details = dict(
+                number=1,
+                offset=0,
+                vtype=None,
+                name=None,
+            )
+
+        stream = self.unlock_volume()
+        yield Volume(
+            fh=stream,
+            size=stream.size,
+            raw=self.fh,
+            vs=self,
+            **volume_details,
+        )
+
+    def unlock_with_volume_encryption_key(self, key: bytes, keyslot: Optional[int] = None) -> None:
+        try:
+            if keyslot is None:
+                for keyslot in self.luks.keyslots.keys():
+                    try:
+                        self.luks.unlock(key, keyslot)
+                        break
+                    except ValueError:
+                        continue
+                else:
+                    raise ValueError("Failed to find matching keyslot for provided volume encryption key")
```

Comment:
Can you also add an error if the LUKS version is 1?

Reply:
I improved the error logging in volume.py to make the existing error more visible.